### ISSUE TITLE
Extremely high CPU usage after file dump to mergerfs pool

### ISSUE THREAD

**Issue by mitzsch** [trapexit/mergerfs#904]
**Describe the bug**
While adding files to the mergerfs array everything is fine. I can see a low CPU usage of about 10 to 20 percent for mergerfs while copying the files from my windows pc over to the linux system (smbd - 10gbit). When the windows file transfer seems to be over there is still drive activity on the server (checked with nmon). RAM flush to disk - I think. When it´s done the mergerfs usage explodes. (1000+% cpu!) without any visible action. After 10 seconds or so everything is back to normal. I have absolutely no idea why. Although It doesn´t really matter on my system (24 cores :D ) I think this is not as expected. Could PLEX be the culprit because of the immediate file access?

**To Reproduce**

Copying files to an array and wait until all files are flushed from RAM to disk. CPU usage for mergerfs explodes.

**Expected behavior**

mergerfs CPU usage should not explode after adding files to the array.

**System information:**
OS:  Ubuntu 20.04.2 LTS x86_64
Kernel: 5.4.0-73-generic
Host: Dell R730xd - Dual Intel Xeon E5-2620 v3 - 32gb RAM
mergerfs version:  2.32.4
Mount options: `/mnt/disk* /mnt/array fuse.mergerfs allow_other,use_ino,cache.files=off,dropcacheonclose=true,category.create=mfs,minfreespace=110G,fsname=mergerfsPool 0 0`
Disks - Filesystem => all are **ext4** formatted + snapraid
```
Filesystem      Size  Used Avail Use% Mounted on
udev             16G     0   16G   0% /dev
tmpfs           3.2G  4.4M  3.2G   1% /run
/dev/sdm2       110G   14G   91G  13% /
tmpfs            16G  4.0K   16G   1% /dev/shm
tmpfs           5.0M  4.0K  5.0M   1% /run/lock
tmpfs            16G     0   16G   0% /sys/fs/cgroup
/dev/sdm1       511M  7.9M  504M   2% /boot/efi
mergerfsPool     27T   19T  8.1T  70% /mnt/array
/dev/nvme0n1p1  458G   72G  363G  17% /mnt/appdata1
/dev/sdk1       1.8T  1.5T  323G  83% /mnt/disk3
/dev/sdl1       1.8T  1.6T  278G  85% /mnt/disk1
/dev/sdh1       1.8T  1.5T  269G  85% /mnt/disk2
/dev/sdj1       1.8T  1.5T  303G  84% /mnt/disk4
/dev/sdi1       1.8T  1.5T  210G  88% /mnt/disk6
/dev/sda1       3.6T  3.4T  247G  94% /mnt/parity1
/dev/sdf1       3.6T  1.6T  1.9T  45% /mnt/disk8
/dev/sdb1       3.6T  3.4T   61G  99% /mnt/parity2
/dev/sdg1       3.6T  3.1T  537G  86% /mnt/disk5
/dev/loop0       56M   56M     0 100% /snap/core18/1988
/dev/sde1       3.6T  3.1T  504G  87% /mnt/disk7
/dev/sdc1       3.6T  1.6T  1.9T  45% /mnt/disk10
/dev/sdd1       3.6T  1.6T  1.9T  45% /mnt/disk9
/dev/loop2       33M   33M     0 100% /snap/snapd/11588
/dev/loop1       56M   56M     0 100% /snap/core18/1997
/dev/loop3       71M   71M     0 100% /snap/lxd/19647
/dev/loop4       33M   33M     0 100% /snap/snapd/11841
/dev/loop5       68M   68M     0 100% /snap/lxd/20326
tmpfs           3.2G     0  3.2G   0% /run/user/0
```

 
**Additional context**

Image of the **htop** output.
![IMG_20210515_194951__01 (Large)](https://user-images.githubusercontent.com/43502191/118374318-afed2c00-b5bb-11eb-8273-ec63d78307df.jpg)

**Comment by trapexit** [comment]
Best if you strace mergerfs while that is happening to see what is happening. Very few things happen that isn't triggered by the kernel making requests to mergerfs.

**Comment by trapexit** [comment]
If it's only happening for a short time then perhaps the kernel was buffering the written data and was flushing it? That level of cpu usage is odd though even if that were happening. there is very little within mergerfs which is cpu heavy. It's an IO bound system.

**Comment by mitzsch** [comment]
Thanks for your fast answer! Next time I dump some stuff to the array I will strace mergerfs.

**Comment by mitzsch** [comment]
>  It's an IO bound system.

Do you mean the drives are to slow to keep up?

**Comment by trapexit** [comment]
I mean that mergerfs generally speaking is spending more time doing IO than doing anything with the CPU. There isn't a lot of compute going on. It's moving data and interacting with filesystems.

**Comment by mitzsch** [comment]
hm okay, then it´s even stranger why there is such a high cpu usage seen. I will give feedback next time I dump stuff to the array. Thanks for your help! :)

**Comment by mitzsch** [comment]
Well, now I´ve added some stuff to my server and interestingly enough everything is fine. The CPU usage still rises but only to about 93-100% in htop and only for 1 or 2 seconds (until everything is dumped to the disks). I also used strace - unfortunately, I have no idea debugging the log output. (never done before - and also the text file is 120mb big, the second one is 3mb) 
Should I attach it?

**Comment by trapexit** [comment]
If you're not able to replicate the behavior I'm not sure the value of digging into it. Ideally the trace would be from the time while the CPU is constantly high. Spikes when busy are expected.

**Comment by mitzsch** [comment]
okay, thanks! Well, then I would consider this as a "one-time thing" (maybe kernel bug?). But I will keep an eye on it and whenever it will happen again, I post a strace log. :)

**Comment by trapexit** [comment]
Sounds good. I'll close this for now but feel free to reopen should something come up.

**Comment by mitzsch** [comment]
Lately, I found something out. It is definitely plex that causes this behavior. Whenever I stop the plex service and copy stuff to the array everything is fine, but when plex is running cpu usage explodes (900+%) after the windows file share is over. If you like I can post a strace log :)

**Comment by trapexit** [comment]
Sure. Is it all the time? Or is it just when it's scanning? Plex shouldn't really be interacting with the filesystem except when scanning and playing. In either case thousands of people including myself use plex with mergerfs and don't see this so it would have to be something bespoke.

**Comment by mitzsch** [comment]
Yes, I tried it multiple times and the issue was always there while plex is scanning for new media. It even happens with the newer ubuntu hwe 5.8 kernel).

**Comment by trapexit** [comment]
Can't replicate. A full scan of my large media collection resulted in Plex Media Scanner using upwards of 40% but as low as 5% and mergerfs peaking at 120% while also under load by dozens of other IO intensive apps. This is on a very down clocked (1.8Ghz) Core i7 3700T.

**Comment by mitzsch** [comment]
On my end, it only happens when I add stuff (the windows file transfer is over, flush to disk is happening) and plex scans for changes at the same time. When I scan for new media while not coping new stuff, everything is just fine.

**Comment by raleeper** [comment]
Sorry for jumping in. Since this only happens when adding new material, could the CPU spike come from Plex generating preview thumbnails? Look at the library advanced options. There is a checkbox for "Enable video preview thumbnails". Extracting and storing the thumbnails only happens for new files, takes CPU for decoding the video to get the thumbnails, and causes disk I/O to save the thumbnails.

**Comment by mitzsch** [comment]
I have the option to generate video preview thumbnails turned on, but the plex is not using that much of the cpu, (only 1 core or so), but mergerfs is using nearly 10 cores. (as you can see in the picture of the htop output)

**Comment by mitzsch** [comment]
@raleeper  Nevertheless, it is something I should test. I will create a new library in plex with video preview thumbnails disabled and only automatic scanning for newly added files turned on. Let´s see if the issue also comes up in this scenario. :)

**Comment by trapexit** [comment]
This is the kind of details I needed to try to replicate. All software has their own behaviors and even the software itself has different behaviors. I'll try to test something later tonight.

**Comment by mitzsch** [comment]
Have you been able to test it? 

I also tried different things on my end. (With new CPUs - e5 2640v4) I tried copying stuff to an internal nvme drive that is not part of the array first and then to the array. Unfortunately, copying the files from the nvme drive to the array shows the exact same extreme cpu usage. 1100%+ With plex stopped everything is just fine. I have absolutely no idea, but all I can say is that samba is not the culprit.......
![Screenshot 2021-07-19 154034](https://user-images.githubusercontent.com/43502191/126805844-9fc164ef-859e-49bf-b80a-c252c705eb55.jpg)
